### ISSUE TITLE
InnoDB + utf8mb4 = Specified key was too long; max key length is 767 bytes

### ISSUE THREAD

**Issue by JasonBarnabe** [vanilla/vanilla#2082]
The MySQL `utf8` character set, despite the name, does not support all of UTF-8. Because it only allows for up to 3 bytes per character, it does not support characters on the supplemental plane, most relevantly: emoji and some CJK characters. The `utf8mb4` character set, on the other hand, allows for up to 4 bytes per character, and so [supports all of UTF-8](http://dev.mysql.com/doc/refman/5.7/en/charset-unicode-utf8mb4.html).

[The maximum index length in InnoDB is 767 bytes](http://dev.mysql.com/doc/refman/5.7/en/innodb-restrictions.html). This means that the maximum size of a column that can fit in an index with `utf8` is 767/3=255 characters, while the maximum for `utf8mb4` is 767/4=191 characters.

As a result, if you try to use Vanilla with the InnoDB engine and the `utf8mb4` character set, there will be errors with the following columns:
- GDN_Flag.ForeignURL
- GDN_Regarding.Type
- GDN_Tag.Name
- GDN_User.Email
- GDN_UserAuthentication.ForeignUserKey
- GDN_UserAuthenticationNonce.Nonce
- GDN_UserMeta.Name

None of the data in these tables in my install required more than 150 characters, so resizing the columns to 150 let me use `utf8mb4`, but now when I run /utility/structure it wants to set them back.

I would like Vanilla to limit column indexes to 191 characters or less to support `utf8mb4` on InnoDB.

**Comment by JasonBarnabe** [comment]
Running `grep -E "varchar\(([2-9][0-9][0-9]|1[9][0-9])\).*(key|unique|primary|index|fulltext)" applications/dashboard/settings/structure.php` it looks like these are the columns with this problem in stage:

- GDN_User.Email (200)
- GDN_UserMeta.Name (255)
- GDN_UserAuthentication.ForeignUserKey (255)
- GDN_UserAuthenticationNonce.Nonce (200)
- GDN_Invitation.Email (200)
- GDN_Tag.Name (255)
- GDN_Tag.FullName (255)
- GDN_Regarding.Type (255)

@linc / @tburry I'd like to get this in for 2.2 to save myself some headaches when I upgrade. I need to know how you want to handle this:

a) Reduce the size of the columns. As mentionned I have no data in my DB where this is a problem, but it's possible others do.
b) Reduce the size of the index. I don't think `Gdn_DatabaseStructure` supports this, so I'd have to hack that a bit.
c) "a" for some columns, "b" for others.

**Comment by linc** [comment]
I do not see an issue with dropping those columns to lower limits. @tburry?

**Comment by linc** [comment]
GDN_User.Email might be a little painful for a utility/update on large sites. Fortunately the rest look fairly painless.

**Comment by tburry** [comment]
We can look into dropping these to maybe 100 and 150 characters. I'm going to need to do some testing on blocking during column length changes though.

**Comment by tburry** [comment]
Also note @JasonBarnabe you can make a plugin to limit column types yourself. Here is an example event handler.

```php
   public function Gdn_MySQLStructure_BeforeSet_Handler($Sender, $Args) {
      if (in_array($Sender->TableName(), array('Discussion', 'Comment'))) {
         $Sender->Column('Body', 'mediumtext', FALSE);
      }
   }
```

**Comment by tburry** [comment]
This was resolved with #2712.
